llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -1502,14 +1668,14 @@ void RISCVInsertVSETVLI::doPRE(MachineBasicBlock &MBB) {
   // we need to prove the value is available at the point we're going
   // to insert the vsetvli at.
   if (AvailableInfo.hasAVLReg()) {
-    const MachineInstr *AVLDefMI = &AvailableInfo.getAVLDefMI();
+    SlotIndex SI = AvailableInfo.getAVLVNInfo()->def;
     // This is an inline dominance check which covers the case of
     // UnavailablePred being the preheader of a loop.
-    if (AVLDefMI->getParent() != UnavailablePred)
+    if (LIS->getMBBFromIndex(SI) != UnavailablePred)
```

Comment:
We won't have LiveIntervals here if `-riscv-vsetvli-after-rvv-regalloc` is set to false, we would need to keep around the old pre-ra path and use getUniqueVRegDef

But what do you think about just removing the flag and pre-ra code paths entirely? It would allow us to simplify things here a good bit, and if we need to reverting should be straightforward since the code changes are only in RISCVInsertVSETVLI.cpp.

I can bring this up for discussion in the next RISC-V sync up call

Reply:
By the way `riscv-vsetvli-after-rvv-regalloc=false` situation, pass manager will invoke LiveIntervals and prepare LIS before into vsetvli pass. it is due to `AU.addRequired<LiveIntervals>();`.

---

I put the flag here for two reasons:

1. It provides a convenient way to compare the pre-RA/post-RA behavior of the vsetvli pass.
2. In case someone encounters a bug, they can disable the flag without modifying the source code.

But the final goal is use post-RA version replace the pre-RA version entirely. Therefore, removing the flag and pre-RA code paths in this patch is good and acceptable.

Reply:
> By the way riscv-vsetvli-after-rvv-regalloc=false situation, pass manager will invoke LiveIntervals and prepare LIS before into vsetvli pass. it is due to AU.addRequired<LiveIntervals>();

Ah that makes sense, sorry for the confusion. So to double check my understanding, we'll always have LiveIntervals but we may or not be in SSA depending on `riscv-vsetvli-after-rvv-regalloc`?

Reply:
Yes. pre-ra and post-ra vsetvli pass major different is SSA form or non-SSA form. It is also why I use the `MRI.isSSA()` to select the pre-ra path inside `needVSETVLPHI` now.